packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -116,13 +139,13 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
 
     const { top, left } = scrollPosition.current!;
 
-    const firstRowIndex = Math.min(
-      Math.max(0, apiRef.current.unstable_getLastMeasuredRowIndex()),
-      getIndexFromScroll(top, rowsMeta.positions),
-    );
+    // Clamp the value because the search may return an index out of bounds.
+    // In the last index, this is not needed because Array.slice doesn't include it.
+    const firstRowIndex = Math.min(getNearestIndexToRender(top), rowsMeta.positions.length - 1);
```

Comment:
Couldn't `exponentialSearch` be responsible for returning a clamped value ?

Reply:
Both search algorithms should return the same values when receiving the same arguments. Besides that we do can fix them to return the value inclusive, then do `+1` for the last index. However, I don't want to touch this logic now because it's critical. This `Math.min` is fixing a bug I discovered when rendering the grid with a very small height and no buffer: https://codesandbox.io/s/datagriddemo-demo-mui-x-forked-g9247r?file=/demo.tsx (scroll all the way to the bottom, no rows is rendered).

```diff
diff --git a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
index b71c895f6..a7be89d11 100644
--- a/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
+++ b/packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx
@@ -27,14 +27,14 @@ export function binarySearch(
   offset: number,
   positions: number[],
   sliceStart = 0,
-  sliceEnd = positions.length,
+  sliceEnd = positions.length - 1,
 ): number {
   if (positions.length <= 0) {
     return -1;
   }
 
   if (sliceStart >= sliceEnd) {
-    return sliceStart;
+    return sliceStart - 1;
   }
 
   const pivot = sliceStart + Math.floor((sliceEnd - sliceStart) / 2);
```

seems to be sufficient to return the value inclusive. We always had this wrong `position.length`.